### PR TITLE
Improve dense vector inner-dot

### PR DESCRIPTION
**Pull request by wongalvis14**
Test results:
No. of locales = 8, n = 1024 ^ 3, compiled with --fast
This implementation: 0.021145 seconds
Original reduction: 0.033186 seconds

No. of locales = 1, n = 1024 ^ 3, compiled with --fast
This implementation: 0.106107 seconds
Original reduction: 0.106456 seconds

But this implementation is slower on small sizes compared to the original reduction

### CODE REVIEW

**Comment by ben-albrecht** [comment]
> Original reduction: 0.033186 seconds

Hey @wongalvis14 - could you compare performance to the implementation as this?

```chpl
  return + reduce(A*B);
```

Those range slices are unnecessary, and are probably introducing some overhead (especially in distributed case)

File: modules/packages/LinearAlgebra.chpl [hunk 1]
```diff
@@ -639,7 +639,39 @@ proc inner(A: [?Adom], B: [?Bdom]) {
   if Adom.size != Bdom.size then
     halt("Mismatched size in inner multiplication");
 
-  return + reduce(A[..]*B[..]);
+  var results: [0..#numLocales] etype = 0;
```
[the remark this answers is not in the document]
Comment (vasslitvinov) [reply modules/packages/LinearAlgebra.chpl]: I think `Locales.domain` is called `LocaleSpace`.

I am concerned about this computation going over all locales because this will hurt the performance when `A` and `B` are not distributed. At the very least, this computation needs to stay local in the known cases such as --local compilation and when `A` and `B` are default rectangular. For known distributions, such as Block and Cyclic, the computation should use `targetLocales` instead of `Locales`. While in the common case the two are the same, I am uneasy about introducing a hidden performance bug noticeable in some runs over lots of locales.

I believe that the computation written as `+ reduce(A*B)` distributes/parallelizes using `A`'s leader, i.e. over `A`'s locales. BTW.

File: modules/packages/LinearAlgebra.chpl [hunk 2]
```diff
@@ -639,7 +639,39 @@ proc inner(A: [?Adom], B: [?Bdom]) {
...
+  
+  coforall l in Locales do on l {
+    var localResult: etype = 0; 
+    const maxThreads = here.maxTaskPar;
```
Comment (vasslitvinov) [inline modules/packages/LinearAlgebra.chpl]: Question to reviewers: are we OK that this computation ignores --dataParTasksPerLocale and dataParIgnoreRunningTasks? They would be honored (hopefully) by the `+ reduce A*B` calculation.

**Comment by vasslitvinov** [comment]
Another case where I expect that this improves performance is when doing a lot of inner-products on small vectors. I also expect the original code to catch up a little, in this scenario, when we switch to the new reduction interface, as that will eliminate free/malloc of reduce-op classes.

File: modules/packages/LinearAlgebra.chpl (hunk 1 above)
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply modules/packages/LinearAlgebra.chpl]: @vasslitvinov - Thanks for pointing out `LocaleSpace`, I never remember that :(

Also, good point on using `targetLocales`.. `Locales` should change to `A.targetLocales()` : https://chapel-lang.org/docs/technotes/subquery.html?highlight=targetlocales#targetlocales

I agree that there probably needs to be a check to whether they're default rectangular (non-distributed or not) I was hoping that we would get there once we run some performance tests and confirmed that that's the case

@wongalvis14 - You updated your PR message with 1 locale. However, it is still different than compiling it with `--local` flag or compiling with a chapel that's built with `CHPL_COMM=none`. Is your 1-locale result above comes from any of these? Or did you simply do `-nl 1` while running the benchmark?

I'd like to do some performance tests myself before this PR is merged.

File: modules/packages/LinearAlgebra.chpl (hunk 2 above)
Comment (e-kayrakli) [reply modules/packages/LinearAlgebra.chpl]: Although I think it is an edge case where you have an asynchronous tasks running in conjunction with this routine, it is definitely a good practice and consistent with the other similar places in the modules.

@wongalvis14 - See here for those variables: https://chapel-lang.org/docs/usingchapel/executing.html#controlling-degree-of-data-parallelism

**Comment by wongalvis14** [comment]
@ben-albrecht @e-kayrakli All performance tests are conducted using reduction without slices, i.e. `  return + reduce(A*B);`, and for the single locale case I've set `$CHPL_COMM=none`

**Comment by wongalvis14** [comment]
I'm thinking there're two options for single locale:

1. Keep this change as it is faster than reduce on large input sizes

2. Discard this change for single locale as it is slower on small input sizes, and single locale users would more often compute on small inputs (this is just my assumption)

If I modify the code (in this diff) within a locale to become a single-locale version, it runs slower than the distributed version (i.e. coforall in all locales), not sure why

Also, if `reduce` uses log-reduction, it's probably the better algorithmic choice. This diff mainly targets the communication between locales and performs as much computation locally as possible, and minimizes communication. And thus, this diff would be less effective on single locales. 

I have a very slight preference for option 2.

**Comment by e-kayrakli** [comment]
I think option 2 is the better one for sure.

As for the PR in general, I just wanted to summarize the things that have not been addressed yet:
- Docstring update as suggested by @ben-albrecht 
- Add a comment to show the elegant implementation, and we don't do it only for performance reasons. I expect the performance of `+ reduce (A*B)` to improve over time. And in an ideal world this function should still be implemented that way, distributed or not.
- A short explanation in the PR message.
- Respect `dataParIgnoreRunningTasks` and `dataParTasksPerLocale` as suggested by @vasslitvinov

File: modules/packages/LinearAlgebra.chpl [hunk 3]
```diff
@@ -631,15 +631,63 @@ private proc _matmatMult(A: [?Adom] ?eltType, B: [?Bdom] eltType)
...
 
+pragma "no doc"
+/* Returns ``true`` if the domain is distributed */
+private proc isDistributed(a) param {
+  return !isSubtype(a.domain.dist.type, DefaultDist);
+}
 
 /* Inner product of 2 vectors. */
-proc inner(A: [?Adom], B: [?Bdom]) {
+proc inner(const ref A: [?Adom], const ref B: [?Bdom]) {
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl]: What was the reason for explicit `const ref` intent here? Was it because compiler was complaining for accesses to them within `coforall`? If so, you can specify intents of the `coforall` specificially:

`coforall l in Locales with (const ref A, const ref B) do`

Comment (wongalvis14) [reply modules/packages/LinearAlgebra.chpl]: No, I did it for the same reasons as declaring consts when possible in C/C++.

Comment (e-kayrakli) [reply modules/packages/LinearAlgebra.chpl]: We discussed this in a meeting, but just to go on record, I am fine with this. It'll make sure that we don't modify the args inside the function. Just wanted to make sure if it was done to work around some issue

**Comment by e-kayrakli** [comment]
Note that the added test in this PR fails. Similarly, the following existing tests also fail due to the same issue:

library/packages/LinearAlgebra/correctness/no-dependencies/no-flags/no-flags                                                                 library/packages/LinearAlgebra/correctness/no-dependencies/correctness                                                                              library/packages/LinearAlgebra/correctness/no-dependencies/implementation-unavailable                                                               library/packages/LinearAlgebra/correctness/no-dependencies/promotion-flattening                                                                      release/examples/primers/LinearAlgebralib

**Comment by e-kayrakli** [comment]
Test status:
- [ ] standard config
- [x] gasnet

File: modules/packages/LinearAlgebra.chpl [hunk 4]
```diff
@@ -631,15 +631,63 @@ private proc _matmatMult(A: [?Adom] ?eltType, B: [?Bdom] eltType)
...
+    result = + reduce (A*B);
+  }
+  else {
+    // Replaces `+ reduce (A*B)` for improved distributed performance
+
+    var localResults: [Locales.domain] eltType = 0;
+
+    coforall l in Locales do on l {
+      const maxThreads = if dataParTasksPerLocale==0 then here.maxTaskPar
+        else dataParTasksPerLocale;
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl]: Could you indent this so that `else` and `then` are aligned

File: modules/packages/LinearAlgebra.chpl [hunk 5]
```diff
@@ -631,15 +631,63 @@ private proc _matmatMult(A: [?Adom] ?eltType, B: [?Bdom] eltType)
...
+      const localDomain = A.localSubdomain();
+      const iterPerThread = divceil(localDomain.size, maxThreads);
+      var localResult: eltType = 0; 
+      var threadResults: [0..#maxThreads] eltType = 0;
+
+      coforall tid in 0..#maxThreads {
+        const startid = localDomain.low + tid * iterPerThread;
+        const temp_endid = startid + iterPerThread - 1;
+        const endid = if localDomain.high < temp_endid then 
+                            localDomain.high else temp_endid;
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl]: I want this `if`, `else`, `then` to be better indented somehow as well. The indentation I suggested above may exceed 80 cols so maybe:
```chapel
const endid = if localDomain.high < temp_endid
              then  localDomain.high else temp_endid;
```

**Comment by e-kayrakli** [comment]
Except for two little style comments, I am ready to merge this in. While you are fixing those, could you update the PR message as well?

File: modules/packages/LinearAlgebra.chpl [hunk 6]
```diff
@@ -631,15 +631,63 @@ private proc _matmatMult(A: [?Adom] ?eltType, B: [?Bdom] eltType)
...
+        var myResult: eltType = 0;
+        local {
+          for ind in startid..endid {
+            myResult += A[ind] * B[ind];
```
Comment (ben-albrecht) [inline modules/packages/LinearAlgebra.chpl]: Local blocks will eventually be deprecated. I'd suggest `localAccess()` instead, e.g.

```suggestion
            myResult += A.localAccess(ind) * B.localAccess(ind);
```

File: modules/packages/LinearAlgebra.chpl [hunk 7]
```diff
@@ -631,15 +631,63 @@ private proc _matmatMult(A: [?Adom] ?eltType, B: [?Bdom] eltType)
...
+      var localResult: eltType = 0; 
+      var threadResults: [0..#maxThreads] eltType = 0;
+
+      coforall tid in 0..#maxThreads {
+        const startid = localDomain.low + tid * iterPerThread;
+        const temp_endid = startid + iterPerThread - 1;
+        const endid = if localDomain.high < temp_endid
+                      then  localDomain.high else temp_endid;
+        var myResult: eltType = 0;
+        local {
```
Comment (ben-albrecht) [inline modules/packages/LinearAlgebra.chpl]: Note that we can remove the `local` block now.

**Comment by wongalvis14** [comment]
@e-kayrakli @ben-albrecht Ready for review
